### PR TITLE
Contracts v2 exchange wrapper

### PR DESCRIPTION
**Pull request by AviKozokin**
Added generic exchange wrapper library

### CODE REVIEW

File: contracts/Exchange/GenericExchangeWrapper.sol [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+            mstore(0x40, and(
+              add(add(end, iszero(add(length, mload(_preBytes)))), 31),
+              not(31) // Round down to the nearest 32 bytes.
+            ))
+        }
+
+        return tempBytes;
+    }
+
+    // function concatStorage(bytes storage _preBytes, bytes memory _postBytes) internal {
```
Comment (svechinsky) [inline contracts/Exchange/GenericExchangeWrapper.sol]: Do we need this huge commented chunk?

File: contracts/Exchange/GenericExchangeWrapper.sol [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+        // avoided for loop because stack too deep
+
+        _preBytes[0] = functionSignature[0];
+        _preBytes[1] = functionSignature[1];
+        _preBytes[2] = functionSignature[2];
+        _preBytes[3] = functionSignature[3];
+        bytes memory packedFunctionCallEncoding = concat(_preBytes, data);
+        // make the call 
+        //emit CallingFunction(packedFunctionCallEncoding);
+        if(!exchange.call(packedFunctionCallEncoding)) {
```
Comment (svechinsky) [inline contracts/Exchange/GenericExchangeWrapper.sol]: Overall makes sense, but if do we use the unconcated bytes for something?
The concat operation seems heavy we could just as well transfer a single byte array?

Comment (AviKozokin) [reply contracts/Exchange/GenericExchangeWrapper.sol]: I forgot to suggest that, sorry.
Yes, indeed there is no need for separation, if it doesnt ease fronentd in any way.
Will change.

File: contracts/Exchange/GenericExchangeWrapper.sol (hunk 1 above)
Comment (AviKozokin) [reply contracts/Exchange/GenericExchangeWrapper.sol]: Not really, it came as part of the Library I used.

**Comment by AviKozokin** [comment]
For testing, I built a Mock environemt: Mock tokens and Mock exchange, and A Mock short contract to simulate import and use of the Library, and make the generic call.
I then ran usecases locally using remix.

**Comment by svechinsky** [comment]
Push the tests as well I think

File: contracts/Exchange/GenericExchangeWrapper.sol (hunk 2 above)
Comment (AviKozokin) [reply contracts/Exchange/GenericExchangeWrapper.sol]: @svechinsky 
This means I will remove concat function (i.e all bytes library), and simply accept one data parameter. Correct me if I'm wrong

**Pull request merged**